src/DiagnosticTools.jl

```diff
@@ -42,3 +42,32 @@ function compute_total_mass!(w,L2MM,hh)
   mul!(w,L2MM,hh)
   sum(w)
 end
+
+"""
+  Full diagnostics for the shallow water equations (mass, vorticity, kinetic energy, potential energy, power)
+"""
+function compute_diagnostics_shallow_water!(model, order, Ω, dΩ, dω, qₖ, wₖ, U, V, R, S, L2MM, H1MM, g, h, u, ϕ, F, mass, vort, kin, pot, pow, step, do_print)
+  mass_i = sum(L2MM*Gridap.FESpaces.get_free_dof_values(h))
```

Comment:
Be careful here ... L2MM*Gridap.FESpaces.get_free_dof_values(h) allocates a whole new temporary vector on each call. We want to avoid this. To this end, we can use `mul!` plus input-argument working array, as it is done, e.g., in the following example: 

https://github.com/gridapapps/GridapGeosciences.jl/blob/master/test/WaveEquationCubedSphereTests.jl#L29

I know this is quite low-level conceptually, but the function `mul!` itself belongs to Julia `LinearAlgebra` standard library.